### PR TITLE
Fix infinite loop.

### LINKED ISSUE #14: Undetermined behavior on periodic timer
Hi,

This is somehow linked with #13  but not totally. I prefer to open a new issue on it's own.
Step to reproduce:
* Start a periodic timer with `repeat time` > `task time to execute`
* Start a periodic timer with `repeat time` < `task time to execute`
* Ignore both guard (this is why i didn't link with #13 has it seems the guard works on his own)
* Drop both timer
* Observe that one timer is still running

```rust
use chrono;
use timer;

fn task(id : u8) {
    println!("{}", id);
    std::thread::sleep(std::time::Duration::from_millis(50));
}

fn main() {
    {
        let timer1 = timer::Timer::new();
        let timer2 = timer::Timer::new();
        let now = chrono::offset::Local::now();
        let repeat1 = chrono::Duration::milliseconds(100); // Long duration. The timer will finish his task before next period
        let repeat2 = chrono::Duration::milliseconds(10); // Short duration. The timer will not finish his task before next period
        timer1.schedule(now, Some(repeat1), || task(1)).ignore();
        timer2.schedule(now, Some(repeat2), || task(2)).ignore();
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
    println!("KILL");
    std::thread::sleep(std::time::Duration::from_millis(500));
    println!("exit now");
}

```
Output is:
```
1
2
2
1
KILL
2
2
2
2
2
2
2
2
2
2
exit now
```

NB:
* I think Having one more `2` (and/or a `1`) after `KILL` can eventually happen if the task is started before droping the timer, but async scheduling call println of main before the one of the task. But here, the timer is in a infinite loop.
* If we do not ignore the guard, the process is good for both, finishing both when `KILL` is printed

### PR DESCRIPTION
**Pull request by wescande**
Fix #14 

**Addendum to the issue:**
Without the fix: If a timer a a huge queue of task (no delay), and we drop the timer, it will execute all of them before exiting, but will not execute any task if he should wait for.
With the fix: The timer will stop executing task as soon as the timer is drop. (running task will finish completion)

### CODE REVIEW

**Comment by wescande** [comment]
Hi !

I edit the PR as I initially made a mistake on mutex lock & condvar. But now it's good.

The second commit is purely cosmetic, has we always exit the inner loop on every schedule, we no longer need it.
I am open to suggestion

**Comment by Yoric** [comment]
Looking at this right now – given the  amount of multitasking I'm doing, do not hesitate to ping me if I suddenly remain quiet. It's going to be because I'm doing way too many things at once.

File: src/lib.rs [hunk 1]
```diff
@@ -848,4 +844,26 @@ mod tests {
             let _  = rx.recv();
         }
     }
+
+    #[test]
+    fn test_repeat_overlap() {
```
Comment (Yoric) [inline src/lib.rs:849]: So what property does this test check, exactly?

File: src/lib.rs [hunk 2]
```diff
@@ -157,58 +157,54 @@ impl <T,E> Scheduler<T,E> where E : Executor<T> {
...
+                                date: sched.date + delta,
+                                data: data,
+                                guard: sched.guard,
+                                repeat: Some(delta),
+                            });
+                        } else {
+                            self.executor.execute(sched.data);
+                        }
+                    }
+                    // We did or skip something, immediately check next schedule
```
Comment (Yoric) [inline src/lib.rs:188]: Nit: `skip` => `skipped`.

File: src/lib.rs [hunk 3]
```diff
@@ -157,58 +157,54 @@ impl <T,E> Scheduler<T,E> where E : Executor<T> {
...
-                let now = Utc::now();
-                if let Some(sched) = self.heap.peek() {
-                    if sched.date > now {
-                        // First item is not ready yet, so we need to
-                        // wait until it is or something happens.
-                        sleep = Sleep::AtMost(sched.date.signed_duration_since(now));
-                        break;
+                    Op::Schedule(sched) => {
+                        self.heap.push(sched);
+                        // Skip sleep as we know we have something to do
```
Comment (Yoric) [inline src/lib.rs:205]: Nit: `know` => `now`.

File: src/lib.rs [hunk 4]
```diff
@@ -848,4 +844,26 @@ mod tests {
...
+            timer.schedule(Utc::now(), Some(Duration::milliseconds(10)), move || {
+                thread::sleep(std::time::Duration::from_millis(30));
+                *called.lock().unwrap() = true;
+            }).ignore();
+        }
+
+        thread::sleep(std::time::Duration::from_millis(150));
+        drop(timer);
+        thread::sleep(std::time::Duration::from_millis(150));
+        assert_eq!(*called.lock().unwrap(), true);
```
Comment (Yoric) [inline src/lib.rs:864]: Nit: I think that adding an assertion message would help understand the test.

**Review by Yoric — COMMENTED** [review]
Patch looks good, thanks ! I have a few nitpicks and I haven't read the test yet, I'll do this asap.

For future reference, could you detail as part of the multiline commit message what scenario you're fixing?

File: src/lib.rs (hunk 1 above)
Comment (wescande) [reply src/lib.rs]: I rename it to be more explicit and I add some comment:
```rust
#[test]                                                                  
fn test_overload() {                                                     
    // Make sure that if timer is overload we still can communicate with 
    // and ask it to stop or schedule an other task.                     
    // An overloaded timer is easily done by scheduling a task longer to 
    // execute than it repeat interval.                                  
```

**Comment by wescande** [comment]
Hi @Yoric ,
Thanks for you feedback. I edit my commit to add multi-line scenario but I am not used to this (I did not fully understand what you mean). I did my best, tell me if it's still not great enough.

**BTW**: With this fix, WaitenerChannel mutex is no longer lock during the full 'execution', but only when looking for some more Op. It is almost never lock. Therefore it seem the channel in timer_base is no longer needed. I locally did some change to try with only one thread for scheduler and no channel. It works really great and I don't see any downside to this.
Are you interested for me to do another PR ?

**Comment by Yoric** [comment]
Apologies for another slow review.

Thanks for the additional comments, I understand much better now! I have left a few nitpicks where I felt that comments could be improved, because I'm annoying like that :)

**Comment by wescande** [comment]
Hi !
I did some change.
Tell me if it's ok.

**Review by Yoric — COMMENTED** [review]
> Hi !
> I did some change.
> Tell me if it's ok.

Sorry it took me so long. I think it's going to be easier if I fix the comments myself.

**Review by Yoric — CHANGES_REQUESTED** [review]
Ok, I'm now happy with it :+1:

Could you fold the 5 patches into a single commit and then we'll land it!

**Comment by wescande** [comment]
I squashed everything into one commit. I like your comment edit, (and you are right, this was quicker).
Thanks.


:heavy_check_mark:  Done.

**Review by Yoric — APPROVED** [review]

**Comment by Yoric** [comment]
And... we're good to go!
Thanks for the pull request and my apologies for not reviewing it earlier. I need to find a better workflow with github.

**Pull request merged**